### ISSUE TITLE
Redux persist dynamic reducer injection goes into paused state

### ISSUE THREAD

**Issue by mobinni** [rt2zz/redux-persist#850]
**Issue:**
We are trying to use redux-persist to persist our state tree with dynamically injected reducers. We use create-react-app and react-router to achieve bundle-splitting on a route by route basis.

I've run into multiple different scenarios where redux-persist breaks and does not persist due to unknown issues. To give you a rudimentary example of the code:

We have a function that creates a root reducer:
```typescript
const asyncReducers: Partial<Reducers> = {};
export const createRootReducer = () => {
  const appReducers: Reducers = {
    ...asyncReducers,
    router: routerReducer,
  };
  return combineReducers(appReducers);
}
``` 
Outside of the function we declare a static map of asynchronous reducers we may inject at some point (in this case a counter bundle will inject counter state onto the store state tree)

```typescript
export type State = {
  readonly counter: CounterState;
  readonly router: RouterState;
} & PersistedState;
```
Our `injectReducer` code is a basic implementation that takes the store and checks the reducer you want to inject against the map and validates if the key is available, if so it won't do anything otherwise it will create a new root reducer and replace the reducer.

```typescript
export const injectReducer = <ReducerState, ReducerKey extends keyof State>(
  store: Store<State>,
  { key, reducer }: ReducerEntry<ReducerState, ReducerKey>,
) => {
  if (typeof asyncReducers[key] !== 'undefined') {
    return;
  }
  asyncReducers[key] = reducer;
  store.replaceReducer(createRootReducer());
};
```

Now according to your documentation wrapping the above code in HMR's `module.hot.accept` callback will solve the issues of persistance breaking when injecting a new reducer. However, because our `Bundle` is controlling this on route load, and the expectation is the reducer be fired and inject the state before the container's first render.

```typescript
  public async componentWillMount() {
    const { component, reducerEntry, rootEpic } = await this.props.bundleWillLoad();
    if (reducerEntry) {
      injectReducer(store, reducerEntry);
    }
    if (rootEpic) {
      injectEpic(rootEpic);
    }
    this.setState({ component });
  }
```

What ends up happening is that when this code calls inject reducer (inside the bundle component) the callback of `module.hot.accept` is never called, but the container tries to bootstrap itself and cannot find `state.counter.counter`.

Here's a working example of the issue for injecting the reducer, it's persisting the router and if you change the url to `/` instead of `/counter` you can see in local storage that it does not update:

https://codesandbox.io/embed/8m4047lz8

I tracked this down in the codebase to the persistor being set to `paused = true`, and for now I hardcoded and overwrite through an action dispatcher that basically it in an unpaused state.

```javascript
 store.dispatch({ type: PERSIST });
```

Basically from what I understand, when you inject a reducer outside of the HMR lifecycle, it seems persistance goes into a paused state because the store's root reducer was changed (??) However I'm not 100% certain on this fact, it took me a while to debug the actual output.

I'm thinking supplying an "UNPAUSE" action wouldn't solve the deeper issue here. However I'm not 100% certain what the deeper issue is exactly, I'd have to dig more through the source-code.

Will try to debug further tomorrow. If you have any idea as to why this may be happening or how to fix it, that would be very much appreciated.

Thank you! 

Tagging who I think could respond to this: @rt2zz

**Comment by marcandrews** [comment]
You must call persist() after calling `store.replaceReducer`.

You can do this conveniently if you expose the persistor on your created store. Then, in `injectReducer`:
```
store.replaceReducer(createRootReducer());
store.persistor.persist();
```

**Comment by mobinni** [comment]
@marcandrews We're using the persist-gate to wrap around the persistor, so on our actual store object, persistor is not available.

**Comment by mobinni** [comment]
That solves my issue, thanks @marcandrews!

**Comment by marcandrews** [comment]
Yes, but `persist()` will handle the `register` and `rehydrate` for you.

**Comment by ygr1k** [comment]
This issue was resolved for me in this ticket:
https://github.com/rt2zz/redux-persist/issues/1028
PR with changes https://github.com/rt2zz/redux-persist/pull/1047/files

**Comment by dotsinspace** [comment]
Guys i am encountring strage behaviour or redux-persist. Only static Reducers which are imported only gets persisted reducers added dynamically on runtime using replaceReducer always fails to get persisted. and even tried @marcandrews way. but nothing works. following is my code.

```
/*
 * IMPORTS
 */
import AsyncStorage from '@react-native-async-storage/async-storage' // Npm: react native file system storage.
import Thunk from 'redux-thunk' // Npm: redux thunk library.
import ExpireReducer from 'redux-persist-expire' // Npm: redux state expire.
import Undoable from 'redux-undo' // Npm: redux store undo redo library.
import ReduxReset from 'redux-reset' // Npm: utility for clearing redux store.
import CreateSagaMiddleware from 'redux-saga' // Npm: redux saga middleware.
import _ from 'underscore' // Npm: utility module.
import { reduxBatch } from '@manaflair/redux-batch' // Npm: batched actions and reducers.
import { applyMiddleware, combineReducers, compose, createStore } from 'redux' // Npm: redux library.
import { persistReducer, persistStore } from 'redux-persist' // Npm: redux persist state library.


/*
 * SIBLINGS
 */
import { Theme } from './Reducers/Theme'
import { System } from './Reducers/System'


/*
 * GLOBALS
 */
let Manager, StoreTree


/*
 * OBJECTS
 */
const UndoableReducer = __reducer => ({ past = [], future = [], ...present } = {}, action) => {
  // Only proceed if reducer has name.
  if (__reducer.name) {
    // Const assignment.
    const _reducerName = __reducer.name.match(/[A-Z][a-z]+/ug)?.join?.('_')?.toUpperCase?.()

    /*
     * Register given reducer for undoable
     * work.
     */
    const _Undoes = Undoable(__reducer, {
      'ignoreInitialState': true,
      'undoType': `${_reducerName}_UNDO`,
      'redoType': `${_reducerName}_REDO`,
      'clearHistoryType': `${_reducerName}_CLEAR`,
      'syncFilter': true,
      'neverSkipReducer': true,
      'debug': false
    })

    // Dispatch states to reducer.
    const _StateUpdates = _Undoes({ past, present, future }, action)

    /*
     * Flat present from given reducer.
     * Because Undoable always return {past, present, future}
     * data structure. causing old incompatibility.
     */
    return {
      ..._StateUpdates.present,
      'past': _StateUpdates.past || [],
      'future': _StateUpdates.future || []
    }
  }

  // Return reducer as is.
  return __reducer
}
const CreateReducer = (__key, __reducer = {}) => {
  /*
   * If store tree is empty than add new
   * elements.
   */
  if (StoreTree && _.isEmpty(StoreTree[__key])) StoreTree[__key] = __reducer
  else if (_.isEmpty(StoreTree)) StoreTree = { Theme, System }

  // Return combined tree.
  return combineReducers(StoreTree)
}
const ConfigureStore = __cb => {
  // Const assignment.
  const _configuration = {
    'key': 'App',
    'storage': AsyncStorage,
    'transforms': [
      /*
       * Create a transformer by passing the reducer key and configuration. Values
       * Shown below are the available configurations with default values
       */
      ExpireReducer('Account', {
        // (Required) Seconds after which store will be expired
        'expireSeconds': 10000 * 60 * 60 * 24 * 20,
        /*
         * (Optional) Use it if you don't want to manually set the time in the reducer i.e. at `persistedAtKey`
         * And want the store to  be automatically expired if the record is not updated in the `expireSeconds` time
         */
        'autoExpire': false
      })
    ],
    'debug': true,
    'writeFailHandler': error => throw error
  }

  // Preserve initial state for not-yet-loaded reducers
  const _SagaMiddleware = CreateSagaMiddleware()
  const _PersistedReducer = persistReducer(_configuration, CreateReducer())

  // Update store to manager.
  Manager = createStore(_PersistedReducer, void 0, compose(applyMiddleware(Thunk, _SagaMiddleware), reduxBatch, ReduxReset()))

  // Persist given manager.
  const _PersistedStore = persistStore(Manager, {}, __cb)

  // Optional: Put the reducer manager on the store so it is easily accessible
  Manager.add = (__key, __reducer) => { Manager.replaceReducer(CreateReducer(__key, __reducer)) }

  // Monkey patch store manager for adding middleware.
  const _nextDispatch = Manager.dispatch
  Manager.dispatch = action => _nextDispatch(action)

  // Update persistor to manager.
  Manager.persistor = _PersistedStore

  // Return persisted store.
  return Manager
}


/*
 * EXPORTS
 */
export default ConfigureStore
export { Manager }

```
